Sage-Bionetworks/challenge-registry
# chore: Upgrade to Yarn 3

- Fixes #588
- Fixes #586

## Notes

- This PR upgrades the development environment to Yarn 2+ ("Yarn Berry").
- Yarn 2+ introduces lock file format v2.
  - Look similar to v1 but best to regenerate the lock file with Yarn 2+ if possible.
- [Motivations and performance benchmarks](https://docs.google.com/presentation/d/18UR44fXblQ7WEg0Sc39lPrXHdWDDxmJgkU3Qi7t9qoY/edit#slide=id.gfe41b02ef2_0_304)
- Yarn 3 CLI is almost identical to Yarn 1 CLI.
  - The only difference relevant to our project is:
    - `yarn install --frozen-lockfile` is replaced by `yarn install --immutable`
    - `yarn bin` is not longer displaying the path to the local `node_modules/.bin` folder (used in `dev-env.sh`).

## Testing

1. Checkout this branch.
2. Rebuild the dev container.
3. Remove the existing `node_modules`.
4. Install the Node.js dependencies: `challenge-install`.
5. Start the challenge registry: `nx serve challenge-registry`.
6. Check that the registry loads successfully in the browser. 

> **Note**
>Repeat Steps 3-4 when returning to a branch that uses Yarn 1.

## TODO before merging

- [ ] Replace dev container image tag by whatever value is used in `main`.
